gadget/update.go

```diff
@@ -301,6 +305,43 @@ func onDiskStructureIsLikelyImplicitSystemDataRole(gadgetVolume *Volume, diskLay
 		numPartsInGadget+1 == numPartsOnDisk
 }
 
+func ensureVolumeEMMCCompatibility(gadgetVolume *Volume, diskVolume *OnDiskVolume) (map[int]*OnDiskStructure, error) {
+	gadgetStructIdxToOnDiskStruct := map[int]*OnDiskStructure{}
+	for _, gs := range gadgetVolume.Structure {
+		// ensure the device node exists
+		emmcNode := fmt.Sprintf("%s%s", diskVolume.Device, gs.Name)
+		if _, err := os.Stat(path.Join(dirs.GlobalRootDir, emmcNode)); err != nil {
+			return nil, fmt.Errorf("emmc disk partition %s is specified, but no such disk: %s",
+				gs.Name, path.Join(dirs.GlobalRootDir, emmcNode))
+		}
+
+		// The eMMC special partition volumes have their own sizes, and we should check
+		// the size declared in the gadget structure is atleast equal to or less than this
+		mmcPart, err := disks.DiskFromDeviceName(emmcNode)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get disk for device %s: %v", emmcNode, err)
+		}
+
+		sz, err := mmcPart.SizeInBytes()
+		if err != nil {
+			return nil, fmt.Errorf("cannot get size of device %s: %v", emmcNode, err)
+		}
+
+		if sz < uint64(gs.Size) {
```

Comment:
Do we support `size` field for emmc schema structures? I do not see that in the spec, and I'm not sure if we should allow them - we do not need them as we are not creating these structures. I would prefer to ban the size if still possible. Note that support for size implies also support for min-size too.

Now we only want to check the size to make sure that the content will fit.

Reply:
So really good point on the Size thing - it took a bit of digging and testing but this also matches some of the feedback I got from Ondrej surrounding this. The whole Size thing is more of an oversight, and I changed behaviour here to disallow setting Size in this case.